osquery/remote/enrollment/plugins/http_enrollment.cpp

```diff
@@ -0,0 +1,91 @@
+#include <osquery/enrollment.h>
+
+#include "osquery/remote/requests.h"
+#include "osquery/remote/transports/http.h"
+#include "osquery/remote/serializers/json.h"
+#include <boost/property_tree/json_parser.hpp>
+
+#define MAX_TRIES 5
+
+namespace osquery {
+
+FLAG(string,
+     enrollment_uri,
+     "Not Specified",
+     "The endpoint for server side client enrollment");
+
+FLAG(string,
+     enrollment_secret,
+     "Not Specified",
+     "The secret to be provided to the endpoint to faciliate enrollment");
+
+FLAG(string,
+     app_id,
+     "Not Specified",
+     "The identification number of the application");
+
+class RemoteEnrollmentPlugin : public EnrollmentPlugin {
+ public:
+  std::string getKey(const PluginRequest& request);
+
+ private:
+  Status enroll();
+  std::string client_key_;
+};
+
+REGISTER(RemoteEnrollmentPlugin, "enrollment", "enrollment");
+
+Status RemoteEnrollmentPlugin::enroll() {
+  auto r = Request<HTTPTransport, JSONSerializer>(FLAGS_enrollment_uri);
+  boost::property_tree::ptree params;
+  params.put<std::string>("secret", FLAGS_enrollment_secret);
+  params.put<std::string>("app_id", FLAGS_app_id);
+  auto stat = r.call(params);
+  if (!stat.ok()) {
+    return stat;
+  }
+  // The call was ok, so store the enrolled key
+  boost::property_tree::ptree recv;
+  stat = r.getResponse(recv);
+  if (!stat.ok()) {
+    return stat;
+  }
+  if (recv.count("enrollment_key") > 0) {
+    client_key_ = recv.get<std::string>("enrollment_key", "");
+    return Status(0, "OK");
+  } else {
+    return Status(1, "No key");
+  }
+}
+
+std::string RemoteEnrollmentPlugin::getKey(const PluginRequest& request) {
```

Comment:
You can pass the a boolean here instead of the request if you're only checking for a forced re-enroll.